Public/Src/Engine/Processes/SharedOpaqueJournal.cs

```diff
@@ -0,0 +1,183 @@
+        /// </summary>
+        public static IEnumerable<string> FindAllJournalFiles(string directory)
+        {
+            return Directory.Exists(directory)
+                ? Directory.EnumerateFiles(directory, "*", SearchOption.TopDirectoryOnly)
+                : CollectionUtilities.EmptyArray<string>();
+        }
+
+        /// <summary>
+        /// Returns all paths recorded in the journal file <paramref name="journalFile"/>.
+        /// </summary>
+        /// <remarks>
+        /// Those paths are expected to be absolute paths of files/directories that were written to by the previous build.
+        ///
+        /// NOTE: this method does not validate the recorded paths in any way.  That means that each returned string may be
+        ///   - an invalid path
+        ///   - a path pointing to an absent file
+        ///   - a path pointing to a file
+        ///   - a path pointing to a directory.
+        ///
+        /// NOTE: the strings in the returned enumerable are neither sorted nor deduplicated.
+        /// </remarks>
+        public static IEnumerable<string> ReadRecordedWritesFromJournal(string journalFile)
+        {
+            using (var fileStream = new FileStream(journalFile, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete))
+            using (var bxlReader = new BuildXLReader(stream: fileStream, debug: false, leaveOpen: false))
+            {
+                string nextString = null;
+                while ((nextString = ReadStringOrNull(bxlReader)) != null)
+                {
+                    yield return nextString;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Same as <see cref="ReadRecordedWritesFromJournal(string)"/> except that all exceptions are wrapped in <see cref="BuildXLException"/>
+        /// </summary>
+        public static string[] ReadRecordedWritesFromJournalWrapExceptions(string journalFile)
+        {
+            try
+            {
+                return ReadRecordedWritesFromJournal(journalFile).ToArray();
+            }
+            catch (Exception e)
+            {
+                throw new BuildXLException("Failed to read from shared opaque journal", e);
+            }
+        }
+
+        /// <summary>
+        /// Records that the file at location <paramref name="path"/> was written to.
+        /// 
+        /// Returns whether the path was recorded or skipped.
+        /// </summary>
+        /// <returns>
+        /// <code>true</code> if <paramref name="path"/> was recorded, <code>false</code> 
+        /// if the path was filtered out because of <see cref="m_rootDirectories"/>.
+        /// </returns>
+        /// <remarks>
+        /// NOT THREAD-SAFE.
+        /// </remarks>
+        public bool RecordFileWrite(AbsolutePath path)
+        {
+            bool isUnderSharedOpaqueDirectory = m_rootDirectories.Any(dir => path.IsWithin(m_pathTable, dir));

```

Comment:
>isUnderSharedOpaqueDirectory [](start = 17, length = 28)

This will record outputs that happen to fall under shared opaques but are not shared opaque outputs. Is there any logic to prevent bxl from deleting those?

Reply:
True.  And there is no logic to prevent BuidXL from deleting those.  But that's how it already is.  Such files are already getting deleted by the scrubbing logic because even today we set magic timestamps on *every* that lives inside a shared opaque directory (even if it is a statically declared one)

Reply:
I don't think that's true. We only flag shared opaques that are identified as such, not all output files

---
In reply to: [334136868](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334136868) [](ancestors = 334136868)

Reply:
#96

Reply:
Just checked. We are using ExecutionResult.SharedDynamicDirectoryWriteAccesses, which only contains shared opaque outputs

---
In reply to: [334137271](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334137271) [](ancestors = 334137271,334136868)

Reply:
So I don't think there is a big problem today with deleting *all* the produced outputs beforehand, because that will only happen to pips that have at least one shared opaque (right?).
But maybe a slightly more complicated solution, but that will accrue to the general 'delayed' deletion that we want to eventually go to could be:
1) after flagging all shared opaques as such, after the pip is done, add a marker to the journal file indicating the pip finished successfully
2) On deletion, if the marker is there, only remove the files that you see in the journal AND have the shared opaque marker
3) On deletion, if you don't see the marker for a given journal file, delete all. The pip failed the last time anyway.

---
In reply to: [334139103](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334139103) [](ancestors = 334139103,334137271,334136868)

Reply:
Here is a [unit test](https://github.com/microsoft/BuildXL/blob/f3be55c8bfa386fb03db4e29a991250755e75c7d/Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/SharedOpaqueDirectoryTests.cs#L1399) that checks that all outputs that fall under a shared opaque directory are marked as shared opaque outputs.

Reply:
Ah, you are right, forgot that we made that change. Indeed both static and dynamic outputs are flagged. That might not be the case for exclusive opaques though, but I don't think that's very impactful

---
In reply to: [334139103](https://github.com/microsoft/BuildXL/pull/1029#discussion_r334139103) [](ancestors = 334139103,334137271,334136868)

Reply:
exactly, outputs in exclusive opaques are not flagged, and that doesn't change with this PR either